src/app/content/highlights/components/CardWrapper.tsx

```diff
@@ -1,30 +1,140 @@
+      setCardsHeights((previous) => new Map(previous).set(id, height === null ? 0 : height));
+    }
+  };
+
+  const getTopOffsetForHighlight = (highlight: Highlight) => {
+    if (topOffsets.has(highlight.id)) {
+      return assertDefined(topOffsets.get(highlight.id), 'this has to be defined');
+    } else {
+      const offset = assertDefined(
+        getHighlightTopOffset(container, highlight),
+        `Couldn't get top offset for highlight with an id: ${highlight.id}`
+      );
+      setTopOffsets((state) => new Map(state).set(highlight.id, offset));
+      return offset;
+    }
+  };
+
+  const updateCardsPositions = React.useCallback(() => {
+    const newPositions: Map<string, number> = new Map();
+
+    let lastVisibleCardPosition = 0;
+    let lastVisibleCardHeight = 0;
+
+    for (const [index, highlight] of highlights.entries()) {
+      const topOffset = getTopOffsetForHighlight(highlight);
+
+      const stackedTopOffset = Math.max(topOffset, lastVisibleCardPosition
+        + lastVisibleCardHeight
+        + (index > 0 ? remsToPx(cardMarginBottom) : 0));
+
+      if (cardsHeights.get(highlight.id)) {
+        lastVisibleCardPosition = stackedTopOffset;
+        lastVisibleCardHeight = cardsHeights.get(highlight.id)!;
+      }
+
+      newPositions.set(highlight.id, stackedTopOffset);
+    }
+
+    setCardsPositions(newPositions);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [highlights, cardsHeights]);
+
+  React.useEffect(() => {
+    if (!focusedHighlight && element.current) {
+      element.current.style.transform = '';
+    }
+  }, [focusedHighlight]);
+
+  React.useEffect(updateCardsPositions, [updateCardsPositions]);
+
+  React.useEffect(() => {
+    if (!focusedHighlight) { return; }
+    const position = cardsPositions.get(focusedHighlight.id);
+    // This may be undefined in case of changing a page when highlight is focused
+    // because highlights will be already cleared and this function will try to run
+    // before page changes.
+    if (!position) { return; }
+    const topOffset = getTopOffsetForHighlight(focusedHighlight);
+
+    if (position > topOffset) {
+      assertNotNull(element.current, 'element.current can\'t be null')
+        .style.transform = `translateY(-${position - topOffset}px)`;
+    }
+
+    // This will be undefined for pendingHighlight
+    if (focusedHighlight.elements[0]) {
+      scrollIntoView(focusedHighlight.elements[0] as HTMLElement);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [focusedHighlight, cardsPositions]);
+
+  return highlights.length
+    ? <div className={className} ref={element}>
+      {highlights.map((highlight, index) => {
+        // Rerender cards on window resize to trigger their height calculations.
+        // It's better to do it here than using hooks like useDebouncedWindowSize for each Card.
+        const key = generateHighlightKey(highlight, width, hasQuery, isTocOpen);
```

Comment:
Won't this make the extra logic run on mobile as well? Maybe you want to pick the generated key or just highlight.id depending on the width?

Reply:
There was an issue when a user has screen width set to some value on which cards are hidden by default, but when he closes the ToC then all cards are displayed on the right side.
This means that cards have `0` height at the beginning, and then when they are displayed they have to update their heights by calling `onHeightChange` - it is not being called automatically because cards were already rendered but they were hidden using CSS.